src/choice_map.cpp

```diff
@@ -0,0 +1,442 @@
+}
+
+Tree ChoiceMap::ExtractTree(ExpandedTreeMask &tree_mask_ext) const {
+  const auto dag_root_id = dag_.GetDAGRootNodeId();
+  Assert(tree_mask_ext.find(dag_root_id) != tree_mask_ext.end(),
+         "DAG Root Id does not exist in ExpandedTreeMask map.");
+  Assert(tree_mask_ext[dag_root_id][AdjacentNode::LeftChild] != NoId,
+         "DAG Root Id has no children in ExpandedTreeMask map.");
+  const auto dag_rootsplit_id = tree_mask_ext[dag_root_id][AdjacentNode::LeftChild];
+
+  std::unordered_map<size_t, bool> visited_left;
+  std::unordered_map<size_t, bool> visited_right;
+  std::unordered_map<size_t, Node::NodePtr> nodes;
+
+  for (const auto &[node_id, adj_node_ids] : tree_mask_ext) {
+    std::ignore = adj_node_ids;
+    visited_left[node_id] = false;
+    visited_right[node_id] = false;
+  }
+
+  size_t node_id_counter = dag_.TaxonCount();
+  auto next_node_id = NoId;
+  auto current_node_id = dag_root_id;
+  // Continue until left and right children of rootsplit node have been visited.
+  nodes[dag_rootsplit_id] = nullptr;
+  while (nodes[dag_rootsplit_id] == nullptr) {
+    // If right branch (and left branch) already visited, join child nodes and return
+    // up the tree.
+    if (visited_right[current_node_id]) {
+      const auto left_child_id =
+          tree_mask_ext[current_node_id][AdjacentNode::LeftChild];
+      const auto right_child_id =
+          tree_mask_ext[current_node_id][AdjacentNode::RightChild];
+      nodes[current_node_id] = Node::Join(nodes.at(left_child_id),
+                                          nodes.at(right_child_id), node_id_counter);
+      node_id_counter++;
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::Parent];
+    }
+    // If left branch already visited, go down the right branch.
+    else if (visited_left[current_node_id]) {
+      visited_right[current_node_id] = true;
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::RightChild];
+    }
+    // If node is a leaf, return up the the tree
+    else if (dag_.IsNodeLeaf(current_node_id)) {
+      nodes[current_node_id] = Node::Leaf(current_node_id);
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::Parent];
+    }
+    // If neither left or right child has been visited, go down the left branch.
+    else {
+      visited_left[current_node_id] = true;
+      next_node_id = tree_mask_ext[current_node_id][AdjacentNode::LeftChild];
+    }
+    Assert(next_node_id != current_node_id, "Node cannot be adjacent to itself.");
+    current_node_id = next_node_id;
+  }
+
+  Node::NodePtr topology = nodes[dag_rootsplit_id];
+  topology->Polish();
+  Assert((nodes.size() + 1) == tree_mask_ext.size(),
+         "Invalid TreeMask-to-Tree: Topology did not span every node in "
+         "the TreeMask.");
+  Tree tree(topology, Tree::BranchLengthVector(nodes.size()));
+
+  return tree;
+}
+
+bool ChoiceMap::TreeIsValid(const Tree &tree, const bool is_quiet) const {
+  std::ofstream devnull("/dev/null");
```

Comment:
Elsewhere (mmapped_matrix) I just used a stringstream that gets thrown away. Look for `dev_null`. Perhaps I prefer that a little bit because it's not making a kernel call. You?

Reply:
You're right, that is cleaner.  Switched it.